components/script/dom/document.rs

```diff
@@ -1285,6 +1279,114 @@ impl Document {
         let mut loader = self.loader.borrow_mut();
         loader.load_async(load, listener)
     }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, JSTraceable, HeapSizeOf)]
+enum EndState {
+    Parsing,
+    BeginningOfTheEnd,
+    FiredDOMContentLoaded,
+    ExecutedAsyncScripts,
+    FiredLoadEvent,
+}
+
+impl Document {
+    pub fn the_end(&self) {
+        assert_eq!(self.end_state.get(), EndState::Parsing);
+
+        // Step 1.
+        self.set_ready_state(DocumentReadyState::Interactive);
+
+        // TODO: Step 2: Pop all the nodes off the stack of open elements.
+
+        self.end_state.set(EndState::BeginningOfTheEnd);
+        self.continue_the_end();
+    }
+
+    pub fn continue_the_end(&self) {
+        if self.end_state.get() == EndState::BeginningOfTheEnd {
+            // Step 3.
+            {
+                let mut deferred_scripts = self.deferred_scripts.borrow_mut();
+                while !deferred_scripts.is_empty() {
+                    // Step 3.1.
+                    if self.script_blocking_stylesheets_count.get() > 0 {
+                        // Spin the event loop. We will be called again by
+                        // XXX when this condition changes.
+                        return;
+                    }
+
+                    {
+                        let (ref element, ref script) = deferred_scripts[0];
+                        let script = match *script {
+                            None => {
+                                // Spin the event loop. We will be called again by
+                                // ScriptContext::response_complete when this
+                                // condition changes.
+                                return;
+                            },
+                            Some(ref script) => script.clone(),
+                        };
+
+                        // Step 3.2.
+                        element.execute(script);
+                    }
+
+                    // Step 3.3.
+                    deferred_scripts.remove(0);
+                }
+            }
+
+            // Step 4.
+            self.dispatch_dom_content_loaded();
+        }
+
+        if self.end_state.get() == EndState::FiredDOMContentLoaded {
```

Comment:
According to the spec we should execute this block after queuing the task in the previous step, so maybe we should actually set end_state above rather than in dispatch_dom_content_loaded?